wakuv2/missing_messages.go

```diff
@@ -0,0 +1,257 @@
+	logger := w.logger.With(
+		zap.Stringer("peerID", missingHistoryRequest.peerID),
+		zap.Strings("contentTopics", missingHistoryRequest.contentTopics[batchFrom:batchTo]),
+		zap.String("pubsubTopic", missingHistoryRequest.pubsubTopic),
+		logutils.WakuMessageTimestamp("from", proto.Int64(missingHistoryRequest.lastChecked.UnixNano())),
+		logutils.WakuMessageTimestamp("to", proto.Int64(now.UnixNano())),
+	)
+
+	result, err := w.storeQueryWithRetry(missingHistoryRequest.ctx, func(ctx context.Context) (*store.Result, error) {
+		return w.node.Store().Query(ctx, store.FilterCriteria{
+			ContentFilter: protocol.NewContentFilter(missingHistoryRequest.pubsubTopic, missingHistoryRequest.contentTopics[batchFrom:batchTo]...),
+			TimeStart:     proto.Int64(missingHistoryRequest.lastChecked.Add(-delay).UnixNano()),
+			TimeEnd:       proto.Int64(now.Add(-delay).UnixNano()),
+		}, store.WithPeer(missingHistoryRequest.peerID), store.WithPaging(false, 100), store.IncludeData(false))
+	}, logger, "retrieving history to check for missing messages")
+	if err != nil {
+		logger.Error("storenode not available", zap.Error(err))
+		return err
+	}
+
+	var missingMessages []pb.MessageHash
+
+	for !result.IsComplete() {
+		for _, mkv := range result.Messages() {
+			hash := pb.ToMessageHash(mkv.MessageHash)
+
+			w.poolMu.Lock()
+			alreadyCached := w.envelopeCache.Has(gethcommon.Hash(hash))
+			w.poolMu.Unlock()
+			if alreadyCached {
+				continue
+			}
+
+			missingMessages = append(missingMessages, hash)
+		}
+
+		result, err = w.storeQueryWithRetry(missingHistoryRequest.ctx, func(ctx context.Context) (*store.Result, error) {
+			if err = result.Next(ctx); err != nil {
+				return nil, err
+			}
+			return result, nil
+		}, logger.With(zap.String("cursor", hex.EncodeToString(result.Cursor()))), "retrieving next page")
+		if err != nil {
+			logger.Error("storenode not available", zap.Error(err))
+			return err
+		}
+	}
+
+	if len(missingMessages) == 0 {
+		// Nothing to do here
+		return nil
+	}
+
+	result, err = w.storeQueryWithRetry(missingHistoryRequest.ctx, func(ctx context.Context) (*store.Result, error) {
+		return w.node.Store().QueryByHash(ctx, missingMessages, store.WithPeer(missingHistoryRequest.peerID), store.WithPaging(false, 100))
+	}, logger, "retrieving missing messages")
+	if err != nil {
+		logger.Error("storenode not available", zap.Error(err))
+		return err
+	}
+
+	for !result.IsComplete() {
+		for _, mkv := range result.Messages() {
+			envelope := protocol.NewEnvelope(mkv.Message, mkv.Message.GetTimestamp(), mkv.GetPubsubTopic())
+			w.logger.Info("received waku2 store message",
+				zap.Stringer("envelopeHash", envelope.Hash()),
+				zap.String("pubsubTopic", mkv.GetPubsubTopic()),
+				zap.Int64p("timestamp", envelope.Message().Timestamp),
+			)
+
+			err = w.OnNewEnvelopes(envelope, common.StoreMessageType, false)
```

Comment:
Wondering if we can have some stats/metrics indicating which filters had more missing messages. This may help as a feedback to filter manager where-in it can decide to use different peer if many messages are being missed.

Reply:
This sounds like a good idea! I'll work on those on a separate PR to not overcomplicate this one